Add kpod import command

umohnani8 opened the pull request:
Imports a tarball and saves it as filesystem image.
One error message change to kpod history
Vendor in latest containers/image

Signed-off-by: Urvashi Mohnani <umohnani@redhat.com>

umohnani8 commented:
@rhatdan PTAL

rhatdan commented on cmd/kpod/import.go:
```diff
@@ -0,0 +1,190 @@
+		Description: importDescription,
+		Flags:       importFlags,
+		Action:      importCmd,
+		ArgsUsage:   "TARBALL",
```
Show me the optional reference.

[ reference ]

rhatdan commented on cmd/kpod/import.go:
```diff
@@ -0,0 +1,190 @@
+			Usage: "Set commit message for the image imported",
+		},
+	}
+	importDescription = "Imports a tarball and saves it as a root filesystem image.\n" +
```
Use the ` 

`
Syntax.

umohnani8 replied:
fixed

rhatdan commented on cmd/kpod/import.go:
```diff
@@ -0,0 +1,190 @@
+	if _, err := url.ParseRequestURI(source); err == nil {
+		file, err := downloadFromURL(source)
+		if err != nil {
+			return nil
```
Return nil or err?

umohnani8 replied:
should be err, fixed

rhatdan commented on cmd/kpod/import.go:
```diff
@@ -0,0 +1,190 @@
+		return errors.Errorf("too many arguments, need 2 only")
+	}
+
+	if _, err := url.ParseRequestURI(source); err == nil {
```
What happens if err != nil?

umohnani8 replied:
fixed

rhatdan commented on libpod/runtime_img.go:
```diff
@@ -877,8 +885,70 @@ func (r *Runtime) GetHistory(image string) ([]ociv1.History, []types.BlobInfo, s
+	if reference == "" {
+		newImg, err := src.NewImage(sc)
+		if err != nil {
+			return err
```
Should be wrapped?

umohnani8 replied:
wrapped.

rhatdan commented on libpod/runtime_img.go:
```diff
@@ -877,8 +885,70 @@ func (r *Runtime) GetHistory(image string) ([]ociv1.History, []types.BlobInfo, s
+
+	policy, err := signature.DefaultPolicy(sc)
+	if err != nil {
+		return err
```
Should this be wrapped?

umohnani8 replied:
wrapped

rhatdan commented on libpod/runtime_img.go:
```diff
@@ -877,8 +885,70 @@ func (r *Runtime) GetHistory(image string) ([]ociv1.History, []types.BlobInfo, s
+
+	policyContext, err := signature.NewPolicyContext(policy)
+	if err != nil {
+		return err
```
Wrapped?

umohnani8 replied:
wrapped

mheon commented:
I think you need to redo vendoring the update c/image. A lot more than just containers/image was updated.

umohnani8 commented:
@mheon I just did containers/image because I needed the new tarball transport that was added. What else should I update in the vendor?

mheon commented:
@umohnani8 The problem is that you picked up too much, not too little - the kubernetes and CNI vendors also updated. I think you need to specifically update only the containers/image vendor with vndr (specify the repo and updated commit directly to vndr). Alternatively, you could directly revert the changes to all the vendors except containers/image by directly manipulating the vendor folder.

umohnani8 commented:
@mheon thant is what I did. I changed the commit id of containers/image to match the tarball commit and just vendored that it. I can try it again.

mheon commented:
@umohnani8 A few of the other vendors (kubernetes, CNI to name a few) don't have specific commits IDs attached, so they have a tendency to update whenever we vendor something new in. I think @runcom and @mrunalp are familiar with how to work around this - might want to ask them for help?

umohnani8 commented:
So when I do `vndr github.com/containers/image` with the tarball commit id, it shows me a bunch of files have been deleted, which include the Kubernetes and CNI files.
I just selected the containers/image files and committed those.

mheon commented:
Vendoring looks fine now, thanks @umohnani8 

I'll do a more thorough review after lunch

umohnani8 commented:
thanks for pointing that out @mheon :)

mheon commented:
Bunch of pull-related tests seem broken. Unclear if this is the new CI acting up or an actual regression.

umohnani8 commented:
@mheon the tests should pass now. Something with the new containers/image was causing them to break.
I am not sure what exactly is happening with containers/image right now, but I just vendored in the changes for the tarball transport to support kpod import. @mrunalp @runcom is that acceptable?

rhatdan commented:
Well I like the fact that you got rid of Nalins pool

mheon commented:
@umohnani8 Containers/image broke a lot of kpod image-related functions, so you dropped the parts that broke us? I'm a bit concerned that we're broken on the upstream code - we should try and figure out how to fix either c/image or how we use c/image so we can use all the updated code.

umohnani8 commented:
@mheon I don't believe I dropped anything that we already had. I just added the tarball folder and changes to the transports folder. There were other changes to containers/image, which was breaking our tests so I didn't commit those after vendoring in containers/image.

mheon commented:
@umohnani8 I'm just a little iffy on cherry picking a partial new commit from c/image. I think we should bother someone on the image side to ask why the new commit broke kpod.

umohnani8 commented:
@mheon the main thing that was breaking was pulling a docker-archive file - it kept throwing the error "blob size mismatch". There was a PR in containers/image that got merged when it shouldn't have and I spoke to @nalind, but he said most of the changes to that was for containers-storage, so weird that docker-archive is messing up.
I can pull in the changes and try to get to the bottom of why its causing pulling from docker-archive to fail.

mheon commented:
@umohnani8 If it's not too much trouble, sure - I'd just like to make sure we aren't going to hit this again next time we have to update c/image.

rhatdan commented:
I agree with Matt, I would rather get this cleaned up in containers/image then get kpod import in.

umohnani8 commented:
All tests should pass now. I was vendoring in from containers/image master and not @nalind's fork, which was causing compression issues with docker-archive therefore breaking the tests.
Vendored in latest from Nalin's fork and all tests should pass now.
@mheon @rhatdan PTAL.

rhatdan commented on cmd/kpod/import.go:
```diff
@@ -0,0 +1,192 @@
+	"strings"
+
+	"github.com/projectatomic/libpod/libpod"
+
```
Remove space

umohnani8 replied:
fixed

rhatdan commented on cmd/kpod/import.go:
```diff
@@ -0,0 +1,192 @@
+	"github.com/projectatomic/libpod/libpod"
+
+	"github.com/opencontainers/image-spec/specs-go/v1"
+
```
remove space

umohnani8 replied:
fixed

rhatdan commented on cmd/kpod/import.go:
```diff
@@ -0,0 +1,192 @@
+	importFlags = []cli.Flag{
+		cli.StringFlag{
+			Name:  "change, c",
+			Usage: "Apply imgspecv1 configurations to the created image",
```
Docker says:
  -c, --change list      Apply Dockerfile instruction to the created image (default [])

I have no idea what a imgspecv1 configurations are.


Man page says
```
       -c, --change=[]
          Apply specified Dockerfile instructions while importing the image
          Supported Dockerfile instructions: CMD|ENTRYPOINT|ENV|EXPOSE|ONBUILD|USER|VOLUME|WORKDIR
```

rhatdan replied:
Change should be a stringsliceflag.  You can specify it multiple times.

umohnani8 replied:
Yeah, I wasn't sure if we wanted to call it docker file instructions. The struct that holds all these configurations is the imgspecv1.Image struct so I used that.
Should I rephrase it to be dockerfile instructions or?

rhatdan replied:
Well instead of saying Dockerfile, why not list the possible options.

umohnani8 replied:
fixed

rhatdan commented on cmd/kpod/import.go:
```diff
@@ -0,0 +1,192 @@
+		},
+		cli.StringFlag{
+			Name:  "message, m",
+			Usage: "Set commit message for the image imported",
```
`Set commit message for imported image`
Reads better

umohnani8 replied:
fixed

rhatdan commented on cmd/kpod/import.go:
```diff
@@ -0,0 +1,192 @@
+			Usage: "Set commit message for the image imported",
+		},
+	}
+	importDescription = `Imports a tarball and saves it as a root filesystem image.
```
Create a container image from the contents of the specified tarball  (.tar, .tar.gz, .tgz, .bzip, .tar.xz, .txz).  
Note remote tar balls can be specified, via web address.

Optionally tag the image.  You can specify the Dockerfile instructions using the --change option.

umohnani8 replied:
fixed

rhatdan commented on cmd/kpod/import.go:
```diff
@@ -0,0 +1,192 @@
+)
+
+func importCmd(c *cli.Context) error {
+	if err := validateFlags(c, historyFlags); err != nil {
```
historyFlags?

umohnani8 replied:
woops, fixed.

rhatdan commented on cmd/kpod/import.go:
```diff
@@ -0,0 +1,192 @@
+	args := c.Args()
+	switch len(args) {
+	case 0:
+		return errors.Errorf("need to give the path to the tarball")
```
Can you specify "-"

"must specify a tarball of '-' for stdin'

umohnani8 replied:
fixed

rhatdan commented on cmd/kpod/import.go:
```diff
@@ -0,0 +1,192 @@
+		source = args[0]
+		opts.Reference = args[1]
+	default:
+		return errors.Errorf("too many arguments, need 2 only")
```
too many arguments.  Usage tarball [reference]

rhatdan commented on cmd/kpod/import.go:
```diff
@@ -0,0 +1,192 @@
+	if err == nil && u.Scheme != "" {
+		file, err := downloadFromURL(source)
+		if err != nil {
+			return err
```
This should be done after all other parsing, so we don't pull an image and then fail later.

umohnani8 replied:
fixed

rhatdan commented on cmd/kpod/import.go:
```diff
@@ -0,0 +1,192 @@
+	}
+
+	changes := v1.ImageConfig{}
+	if c.IsSet("change") {
```
Change is a slice, should be a for loop.

umohnani8 replied:
fixed

rhatdan commented on docs/kpod-import.1.md:
```diff
@@ -0,0 +1,87 @@
+## OPTIONS
+
+**--change, -c**
+Apply imgspecv1 configurations to the created image
```
Dockerfile configurations.

rhatdan replied:
Can be specified multiple times.

umohnani8 replied:
fixed

rhatdan commented on libpod/runtime_img.go:
```diff
@@ -1080,3 +1121,32 @@ func findImageInSlice(images []storage.Image, ref string) (storage.Image, error)
 	return storage.Image{}, errors.New("could not find image")
 }
+
+// getIMgeDigest creates an image object and uses the hex value of the digest as the image ID
```
Spelling.

umohnani8 replied:
fixed.

rhatdan commented on test/kpod_import.bats:
```diff
@@ -0,0 +1,141 @@
+    cleanup_pods
+    stop_crio
+    rm -f container.tar
+}
```
newline

umohnani8 replied:
fixed.

rhatdan commented:
One last nit, LGTM

@mheon @mitrmac PTAL

mheon commented:
@umohnani8 The code from Nalin's c/image fork is going to land upstream at some point, right? I don't want to be permanently tied to his fork instead of upstream

mheon commented on cmd/kpod/import.go:
```diff
@@ -0,0 +1,190 @@
+
+	runtime, err := getRuntime(c)
+	if err != nil {
+		return errors.Wrapf(err, "Could not get runtime")
```
Errors should start with a lowercase letter.

@rhatdan Golint should have caught this - are we running it in the tests yet?

umohnani8 replied:
fixed.

umohnani8 commented:
@mheon I believe so. Also Nalin's fork is what has been vendored in for cri-o.

mheon commented on cmd/kpod/save.go:
```diff
@@ -91,6 +91,7 @@ func saveCmd(c *cli.Context) error {
 	for _, image := range args {
 		dest := dst + ":" + image
 		if err := runtime.PushImage(image, dest, saveOpts); err != nil {
+			os.Remove(output)
```
This can return an error, should check that

umohnani8 replied:
fixed

mheon commented on libpod/runtime_img.go:
```diff
@@ -877,8 +867,59 @@ func (r *Runtime) GetHistory(image string) ([]ociv1.History, []types.BlobInfo, s
+
+	sc := common.GetSystemContext("", "")
+
+	// 	if reference not given, get the image digest
```
Lot of extra whitespace in this comment

umohnani8 replied:
fixed

mheon commented on libpod/runtime_img.go:
```diff
@@ -877,8 +867,59 @@ func (r *Runtime) GetHistory(image string) ([]ociv1.History, []types.BlobInfo, s
+	}
+
+	annotations := make(map[string]string)
+	annotations[ociv1.AnnotationDescription] = "test image built"
```
Should this be kept in? Looks like a debugging statement

umohnani8 replied:
removed.

mheon commented on libpod/runtime_img.go:
```diff
@@ -1080,3 +1121,32 @@ func findImageInSlice(images []storage.Image, ref string) (storage.Image, error)
+	return "@" + digest.Hex(), nil
+}
+
+func getPolicyContext(ctx *types.SystemContext) (*signature.PolicyContext, error) {
```
Can we get a brief comment on this function?

umohnani8 replied:
fixed

umohnani8 commented:
@rhatdan tests passed.

rhatdan commented on cmd/kpod/save.go:
```diff
@@ -91,6 +91,9 @@ func saveCmd(c *cli.Context) error {
 	for _, image := range args {
 		dest := dst + ":" + image
 		if err := runtime.PushImage(image, dest, saveOpts); err != nil {
+			if err2 := os.Remove(output); err2 != nil {
```
I think you would just logrus this.  But return the original error.

umohnani8 replied:
fixed

rhatdan commented:
@mheon PTAL
LGTM other then the comment above.

mheon commented:
LGTM. @umohnani8 Can you add an issue after this merges to switch back to containers/image master once Nalin's changes merge?

rhatdan commented:
@rh-atomic-bot r+
